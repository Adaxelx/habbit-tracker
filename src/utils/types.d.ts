export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: boolean;
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}
