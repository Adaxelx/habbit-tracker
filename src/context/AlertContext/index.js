export { default, useAlertContext } from './AlertContext';
