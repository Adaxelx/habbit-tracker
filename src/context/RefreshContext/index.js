export { default, useRefreshContext } from './RefreshContext';
