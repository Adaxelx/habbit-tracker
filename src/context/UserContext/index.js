export { default, useUserContext, UserContext } from './UserContext';
