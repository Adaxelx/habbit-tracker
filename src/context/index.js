export { default as ThemeProvider, useThemeContext } from './ThemeContext';
export { default as UserProvider, useUserContext, UserContext } from './UserContext';
export { default as AlertProvider, useAlertContext, AlertContext } from './AlertContext';
export { default as RefreshProvider, useRefreshContext } from './RefreshContext';
