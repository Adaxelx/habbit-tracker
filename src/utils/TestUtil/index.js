export { default } from './TestUtil.ts';
