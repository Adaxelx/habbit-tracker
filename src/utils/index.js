export { default as TestUtil } from './TestUtil';
export { default as RenderWithRouter } from './RenderWithRouter';
export * from './validators';
export * from './types.d';
