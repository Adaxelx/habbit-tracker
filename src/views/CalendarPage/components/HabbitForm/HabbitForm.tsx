import React, { useRef, useEffect, useCallback } from 'react';
import { PopUp, Input, Button, DateInput, Select } from 'components';
import { FieldError, useForm } from 'react-hook-form';
import {
  EventSend,
  createRestrictedLengthObject,
  AlertTypes,
  FormHabbit,
  createTimeValidation,
} from 'utils';
import { useAlertContext, useUserContext, useRefreshContext } from 'context';
import { postEvent } from 'views/CalendarPage/CalendarPage.api';
// import { getISODate } from 'constants/calendar';
import { StyledWrapper } from './HabbitForm.css';
import { WeekDaysInput } from '..';

const { SUCCESS } = AlertTypes;

const HabbitForm = ({ handleClose, open, labels, event }: FormHabbit) => {
  // const dateStartDefault = new Date();
  const { register, handleSubmit, errors, control, reset, watch } = useForm<EventSend>({
    defaultValues: {
      timeStart: '00:10',
      timeEnd: '23:50',
      label: '',
      daysOfWeek: [],
      dateStart: '',
      dateEnd: '',
    },
  });

  const alertC = useRef(useAlertContext());

  const { token } = useUserContext();
  const { handleRefHabbit } = useRefreshContext();

  useEffect(() => {
    let defaultValues = {};
    if (event) {
      const { label, dateStart, dateEnd } = event;
      defaultValues = {
        ...event,
        label: label?._id,
        dateStart: new Date(dateStart),
        dateEnd: new Date(dateEnd),
      };
      reset(defaultValues);
    }
  }, [event]);

  const onSubmit = async (data: EventSend) => {
    try {
      await postEvent(token, data, event?._id);
      alertC.current.showAlert(`Succesfuly ${event?._id ? 'edited' : 'added'} habbit.`, SUCCESS);
      handleClose();
      handleRefHabbit();
    } catch (err) {
      alertC.current.showAlert(err.message);
    }
  };
  const dateStart = watch('dateStart') ? new Date(watch('dateStart')) : undefined;
  const dateEnd = watch('dateEnd') ? new Date(watch('dateEnd')) : undefined;

  const datePlus7 = useCallback((datePassed, diff) => {
    if (datePassed) {
      const date = new Date(datePassed.getTime());
      date.setDate(date.getDate() + diff);
      return date;
    }
    return undefined;
  }, []);

  return (
    <PopUp open={open} handleClose={handleClose} header="Add habbit">
      <StyledWrapper as="form" onSubmit={handleSubmit(onSubmit)} noValidate>
        <Input
          name="title"
          label="Title"
          id="title"
          refVal={register(createRestrictedLengthObject('title'))}
          minLength={6}
          maxLength={30}
          error={errors.title}
          data-testid="title"
        />
        <Input
          name="description"
          label="Description"
          id="description"
          refVal={register()}
          error={errors.description}
          data-testid="description"
          type="textarea"
        />
        <WeekDaysInput control={control} error={(errors.daysOfWeek as unknown) as FieldError} />
        <DateInput
          control={control}
          name="dateStart"
          header="Date start"
          maxDate={datePlus7(dateEnd, -7)}
          error={errors.dateStart}
        />
        <DateInput
          control={control}
          name="dateEnd"
          header="Date end"
          minDate={datePlus7(dateStart, 7)}
          error={errors.dateEnd}
        />
        <Input
          name="timeStart"
          label="Time start"
          id="timeStart"
          refVal={register(createTimeValidation('00:00', watch('timeEnd')))}
          error={errors.timeStart}
          type="time"
          data-testid="timeStart"
        />
        <Input
          name="timeEnd"
          label="Time end"
          id="timeEnd"
          refVal={register(createTimeValidation(watch('timeStart'), '23:59'))}
          error={errors.timeStart}
          type="time"
          data-testid="timeEnd"
        />
        <Select
          name="label"
          label="Label"
          control={control}
          options={labels?.map(({ _id, title, ...rest }) => ({ key: _id, value: title, ...rest }))}
        />
        <Button size="s" mt="16px" type="submit" data-testid="submit">
          Send
        </Button>
      </StyledWrapper>
    </PopUp>
  );
};

export default HabbitForm;
