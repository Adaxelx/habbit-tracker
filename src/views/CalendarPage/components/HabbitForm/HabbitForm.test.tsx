import React from 'react';
import 'jest-styled-components';
import { waitFor } from '@testing-library/dom';
import { AlertContext } from 'context';
import { AlertTypes, FormHabbit, TestUtil } from 'utils';
import HabbitForm from './HabbitForm';

describe('HabbitForm', () => {
  let util: TestUtil;
  let props: FormHabbit;
  let fail: boolean;
  const alertC = { showAlert: jest.fn() };

  const mockedFetch = () => {
    if (fail) {
      return Promise.resolve({
        status: 500,
        ok: false,
        json: () => Promise.resolve({ message: 'error' }),
      });
    }

    return Promise.resolve({
      status: 201,
      ok: true,
    });
  };

  beforeAll(() => {
    jest.spyOn(window, 'fetch');
  });

  beforeEach(() => {
    props = {
      handleClose: jest.fn(),
      open: true,
      labels: [{ _id: '1', color: '#ffffff', title: 'label', userId: '123' }],
    };
    fail = false;
    jest.clearAllMocks();
    window.fetch.mockImplementation(mockedFetch);
  });

  it('should match snapshot', () => {
    util = new TestUtil(
      (
        <AlertContext.Provider value={alertC}>
          <HabbitForm {...props} />
        </AlertContext.Provider>
      ),
    );
    expect(util.render.asFragment()).toMatchSnapshot();
  });

  it('should send habbit', async () => {
    util = new TestUtil(
      (
        <AlertContext.Provider value={alertC}>
          <HabbitForm {...props} />
        </AlertContext.Provider>
      ),
    );

    util.setValue('title', 'title123');
    util.setValue('description', 'description');
    util.click('Mon');
    util.click('Mon');
    util.click('Tue');
    util.click('Mon');
    util.setValueByL('Date start', new Date('2021-06-06'));
    util.setValueByL('Date end', new Date('2021-06-16'));
    util.setValue('timeStart', '12:15');
    util.setValue('timeEnd', '15:15');
    util.setValue('label', '1');
    util.click('submit');

    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('Succesfuly added habbit.', AlertTypes.SUCCESS);
  });

  it('should not send habbit if api fails', async () => {
    fail = true;

    util = new TestUtil(
      (
        <AlertContext.Provider value={alertC}>
          <HabbitForm {...props} />
        </AlertContext.Provider>
      ),
    );

    util.setValue('title', 'title123');
    util.setValue('description', 'description');
    util.click('Mon');
    util.setValueByL('Date start', new Date('2021-06-06'));
    util.setValueByL('Date end', new Date('2021-06-16'));
    util.setValue('timeStart', '12:15');
    util.setValue('timeEnd', '15:15');
    util.setValue('label', '1');
    util.click('submit');

    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('error');
  });

  it('should send edited habbit', async () => {
    props = {
      ...props,
      event: {
        _id: '123',
        userId: '1233',
        title: 'title123',
        description: 'description',
        daysOfWeek: [0],
        dateStart: '2021-06-06',
        dateEnd: '2021-06-16',
        timeStart: '12:15',
        timeEnd: '15:15',
        label: { _id: '1', color: '#ffffff', title: 'xd', userId: 'xd' },
      },
    };

    util = new TestUtil(
      (
        <AlertContext.Provider value={alertC}>
          <HabbitForm {...props} />
        </AlertContext.Provider>
      ),
    );

    util.setValue('title', 'title1234343');
    util.click('submit');

    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('Succesfuly edited habbit.', AlertTypes.SUCCESS);
  });
});
