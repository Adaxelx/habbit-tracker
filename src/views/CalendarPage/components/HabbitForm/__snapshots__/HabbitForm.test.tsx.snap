// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HabbitForm should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  height: 100vh;
  position: fixed;
  top: 0;
  left: 0;
  z-index: 1000;
  background-color: rgba(0,0,0,0.3);
}

<div
    class="c0"
  />
  .c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 1rem;
  width: 100%;
  max-width: 300px;
}

.c7 {
  font-size: 1.5rem;
  margin-bottom: 0.75rem;
}

.c8 {
  font-size: 1.25rem;
  width: 100%;
  max-width: 300px;
}

.c8:focus {
  outline: thin dotted;
}

.c9 {
  font-size: 1.25rem;
  width: 100%;
  max-width: 300px;
}

.c9:focus {
  outline: thin dotted;
}

.c10 {
  width: 100%;
  display: grid;
  grid-template-columns: repeat(4,1fr);
  grid-template-rows: repeat(2,1fr);
}

.c11 {
  background-color: #FFFFFF;
  border: 1px solid black;
  height: 50px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 1.5rem;
}

.c5 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c14 {
  color: #000000;
  background-color: #FFC136;
  border: 1px solid #707070;
  padding: 0.75rem;
  font-size: 1rem;
  min-width: 150px;
  display: inline-block;
  text-align: center;
  margin-top: 16px;
  margin-left: 0;
  margin-right: 0;
}

.c14:hover {
  cursor: pointer;
}

.c14:disabled {
  -webkit-filter: grayscale(1);
  filter: grayscale(1);
}

.c0 {
  position: fixed;
  width: calc(100% - 1rem - 1rem);
  height: calc(90vh - 1rem - 1rem);
  top: 1rem;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  padding: 1rem 0;
  background-color: #AFE9FF;
  border: 1px solid #707070;
  z-index: 1000;
  max-width: 350px;
  overflow: scroll;
}

.c4 {
  width: 100%;
  text-align: center;
  font-size: 2rem;
}

.c1 {
  width: 50px;
  height: 50px;
  position: relative;
  display: block;
  margin-left: auto;
  margin-right: 0;
  border: none;
  background-color: transparent;
}

.c1:hover {
  cursor: pointer;
}

.c1:disabled {
  -webkit-filter: grayscale(1);
  filter: grayscale(1);
}

.c2 {
  height: 100%;
  width: 2px;
  position: absolute;
  top: 0%;
  left: 50%;
  background-color: #FF9191;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c3 {
  height: 100%;
  width: 2px;
  position: absolute;
  top: 0%;
  left: 50%;
  background-color: #FF9191;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
}

.c12 {
  width: 100%;
  max-width: 300px;
  margin-bottom: 1rem;
}

.c13 {
  width: 100%;
  max-width: 300px;
  font-size: 1.25rem;
}

<div
    class="c0"
  >
    <button
      class="c1"
      data-testid="Add habbit-close"
      type="button"
    >
      <div
        class="c2"
        rotate="45"
      />
      <div
        class="c3"
        rotate="-45"
      />
    </button>
    <h2
      class="c4"
      data-testid="Add habbit-header"
    >
      Add habbit
    </h2>
    <form
      class="c5"
      novalidate=""
    >
      <div
        class="c6"
      >
        <label
          class="c7"
          for="title"
        >
          Title
        </label>
        <input
          class="c8"
          data-testid="title"
          id="title"
          maxlength="30"
          minlength="6"
          name="title"
          type="text"
        />
      </div>
      <div
        class="c6"
      >
        <label
          class="c7"
          for="description"
        >
          Description
        </label>
        <textarea
          class="c9"
          data-testid="description"
          id="description"
          name="description"
        />
      </div>
      <div
        class="c6"
      >
        <label
          class="c7"
        >
          Week days
        </label>
        <div
          class="c10"
        >
          <div
            class="c11"
            data-testid="Mon"
          >
            Mon
          </div>
          <div
            class="c11"
            data-testid="Tue"
          >
            Tue
          </div>
          <div
            class="c11"
            data-testid="Wed"
          >
            Wed
          </div>
          <div
            class="c11"
            data-testid="Thu"
          >
            Thu
          </div>
          <div
            class="c11"
            data-testid="Fri"
          >
            Fri
          </div>
          <div
            class="c11"
            data-testid="Sat"
          >
            Sat
          </div>
          <div
            class="c11"
            data-testid="Sun"
          >
            Sun
          </div>
        </div>
      </div>
      <div
        class="c12"
      >
        <label
          class="c7"
          for="dateStart"
        >
          Date start
        </label>
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <input
              class=""
              id="dateStart"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="c12"
      >
        <label
          class="c7"
          for="dateEnd"
        >
          Date end
        </label>
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <input
              class=""
              id="dateEnd"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="c6"
      >
        <label
          class="c7"
          for="timeStart"
        >
          Time start
        </label>
        <input
          class="c8"
          data-testid="timeStart"
          id="timeStart"
          name="timeStart"
          type="time"
        />
      </div>
      <div
        class="c6"
      >
        <label
          class="c7"
          for="timeEnd"
        >
          Time end
        </label>
        <input
          class="c8"
          data-testid="timeEnd"
          id="timeEnd"
          name="timeEnd"
          type="time"
        />
      </div>
      <div
        class="c6"
      >
        <label
          class="c7"
        >
          Label
        </label>
        <select
          class="c13"
          data-testid="label"
        >
          <option
            value=""
          >
            ----Brak labelu----
          </option>
          <option
            value="1"
          >
            label
          </option>
        </select>
      </div>
      <button
        class="c14"
        data-testid="submit"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</DocumentFragment>
`;
